rfc/src/rfcs/0002-function-stubbing.md

```diff
@@ -0,0 +1,340 @@
+
+Using stubbing, we can specify that the function `rand::random` should be replaced with a mocked version:
+
+```rust
+#[cfg(kani)]
+fn mock_random<T: kani::Arbitrary>() -> T {
+    kani::any()
+}
+
+#[cfg(kani)]
+#[kani::proof]
+#[kani::stub_by("rand::random", "mock_random")]
+fn random_cannot_be_zero() {
+    assert_ne!(rand::random::<u32>(), 0);
+}
+```
+
+Under this substitution, Kani has a single check, which proves that the assertion can fail. Verification time is 0.02 seconds.
+
+### Mocking System Calls
+
+**TODO**
+
+### Mocking Deserialization
+
+**TODO**
+
+## User Experience
+
+This feature is currently limited to stubbing functions and methods.
+We anticipate that the annotations we propose here could also be used when stubbing types, although the underlying technical approach might have to change.
+
+Stubs will be specified per harness; that is, different harnesses can use different stubs (the reasoning being that users might want to mock different behavior for different harnesses).
+Users will specify stubs by attaching the `#[kani::stub_by("<original>", "<replacement>")]` attribute to each harness function.
+The attribute may be specified multiple times per harness, so that multiple (non-conflicting) stub pairings are supported.
+The arguments `original` and `replacement` give the names of functions, relative to the crate of the harness (*not* relative to the module of the harness).
+
+For example, this code specifies that the function `mock_random` should be used in place of the function `rand::random` and the function `my_mod::foo` should be used in place of the function `my_mod::bar` for the harness `my_mod::my_harness`:
+
+```rust
+#[cfg(kani)]
+fn mock_random<T: kani::Arbitrary>() -> T {
+    kani::any()
+}
+
+mod my_mod {
+
+    fn foo(x: u32) -> u32 { ... }
+
+    fn bar(x: u32) -> u32 { ... }
+
+    #[cfg(kani)]
+    #[kani::proof]
+    #[kani::stub_by("rand::random", "mock_random")]
+    #[kani::stub_by("my_mod::foo", "my_mod::bar")]
+    fn my_harness() { ... }
+
+}
+```
+
+### Stub Sets
+
+As a convenience, users will also be able to specify sets of stubs that can be applied to multiple harnesses.
+First, users write a "dummy" function with the name of the stub set, annotate it with the `#[kani::stub_set]` attribute, and add the desired stub pairings as further attributes:
+
+```rust
+#[cfg(kani)]
+#[kani::stub_set]
+#[kani::stub_by("std::fs::read", "my_read")]
+#[kani::stub_by("std::fs::write", "my_write")]
+fn my_io_stubs() {}
+```
+
+When declaring a harness, users can use the `#[kani::use_stub_set("<stub_set_name>")]` attribute to apply the stub set:
+
+```rust
+#[cfg(kani)]
+#[kani::proof]
+#[kani::use_stub_set("my_io_stubs")]
+fn my_harness() { ... }
+```
+
+The same mechanism can be used to aggregate stub sets:
+
+```rust
+#[cfg(kani)]
+#[kani::stub_set]
+#[kani::use_stub_set("my_io_stubs")]
+#[kani::use_stub_set("other_stub_set")]
+fn all_my_stubs() {}
+```
+
+### Accessing Private Fields in Method Stubs
```

Comment:
Part of me would prefer if we leave this out of the initial RFC and just leave it as part of "Open Questions".

Reply:
That'd be fine with me. @adpaco-aws what do you think?

Reply:
Agree, this is a subtle detail that appears way too early in the RFC.

Reply:
I removed this section.